### PR TITLE
Implemented analysis helper class for qtests

### PR DESCRIPTION
**Pull request by tcoates3**
BEGINRELEASENOTES
- Implemented an AnalysisHelper class to compute 
   - mean, rms (TGraph, TH1)
   - mean 90, rms 90 (TGraph, TH1)
- Renamed MeanWithinExpectedTest to PropertyWithinExpectedTest with expanded functionalities
    - Test if property is within-range, higher-than, and lower-than
    - Possible properties are `Mean`, `RMS`, `Median`
    - For `Mean` and `RMS` can specifiy percentage interval
    - Implemented unit test
- Modified GenerateData.C to create TGraphs for testing, and updated test_samples.root to contain these
- Modified test_samples.xml to have examples of all newly-implemented features

ENDRELEASENOTES

### CODE REVIEW

File: source/include/dqm4hep/AnalysisHelper.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+#include <TH1.h>
+#include <TMath.h>
+
+namespace dqm4hep {
+
+  namespace core {
+
+    /** analysis helper class
+     */
+    class AnalysisHelper
```
Comment (rete) [inline source/include/dqm4hep/AnalysisHelper.h:49]: Here you should have separate functions for mean, mean90, rms90, etc ...
Simply call them `AnalysisHelper::mean` or `AnalysisHelper::mean90` and return the value

File: include/dqm4hep/AnalysisHelper.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
+/// \file AnalysislHelper.h
```
Comment (rete) [inline include/dqm4hep/AnalysisHelper.h]: This file should not be there. It's a file that is installed after `make install`

File: source/include/dqm4hep/StatusCodes.h [hunk 3]
```diff
@@ -36,6 +36,19 @@
 #include <execinfo.h>
 #endif
 
+#define SOFT_RETURN_RESULT_IF(Command)                                                                                \
```
Comment (rete) [inline source/include/dqm4hep/StatusCodes.h]: For the use you make, you don't need this macro. Use the `RETURN_RESULT_IF_AND_IF` macro as:
```cpp
RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, operation);
```

File: source/src/AnalysisHelper.cc [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    //----------------------------------------------------------------------------------------------------
+
+    float AnalysisHelper::mainHelper(MonitorElementPtr pMonitorElement, std::string testType, float percentage)
+    {
+
+      if (nullptr == pMonitorElement) {
+        throw StatusCodeException(STATUS_CODE_INVALID_PTR);
+      }
+
+      TH1 *h = pMonitorElement->objectTo<TH1>();
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: The `AnalysisHelper::mean` function should also work for other object such as TGraph (on the y axis)

File: source/src/plugins/MeanWithinExpectedTest.cc [hunk 5]
```diff
@@ -33,6 +33,8 @@
 #include <dqm4hep/QualityTest.h>
 #include <dqm4hep/StatusCodes.h>
 #include <dqm4hep/XmlHelper.h>
+#include <dqm4hep/AnalysisHelper.h>
+//#include <dqm4hep/AnalysisHelper.cc>
```
Comment (rete) [inline source/src/plugins/MeanWithinExpectedTest.cc]: Remove this comment

File: source/src/plugins/MeanWithinExpectedTest.cc [hunk 6]
```diff
@@ -33,6 +33,8 @@
 #include <dqm4hep/QualityTest.h>
 #include <dqm4hep/StatusCodes.h>
 #include <dqm4hep/XmlHelper.h>
+#include <dqm4hep/AnalysisHelper.h>
```
Comment (rete) [inline source/src/plugins/MeanWithinExpectedTest.cc]: The file `MeanWithinExpectedTest.cc` and the test name should be renamed to PropertyWithinExpectedTest(.cc) as it is no longer the mean you can test

File: source/src/plugins/MeanWithinExpectedTest.cc [hunk 7]
```diff
@@ -60,6 +62,10 @@ namespace dqm4hep {
       float m_expectedMean;
       float m_meanDeviationLower;
       float m_meanDeviationUpper;
+      float m_percentage;
+
+      std::vector<std::string> m_testTypes;
```
Comment (rete) [inline source/src/plugins/MeanWithinExpectedTest.cc]: Rename it to `m_properties`

File: source/src/plugins/MeanWithinExpectedTest.cc [hunk 8]
```diff
@@ -87,15 +95,19 @@ namespace dqm4hep {
...
 
-      RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=,
-                       XmlHelper::readParameter(xmlHandle, "MeanDeviationLower", m_meanDeviationLower,
-                                                [this](const float &value) { return value < this->m_expectedMean; }));
+      SOFT_RETURN_RESULT_IF(XmlHelper::readParameter(xmlHandle, "Percentage", m_percentage));
 
-      RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=,
-                       XmlHelper::readParameter(xmlHandle, "MeanDeviationUpper", m_meanDeviationUpper,
-                                                [this](const float &value) { return value > this->m_expectedMean; }));
+      SOFT_RETURN_RESULT_IF(XmlHelper::readParameter(xmlHandle, "MeanDeviationLower", m_meanDeviationLower));
```
Comment (rete) [inline source/src/plugins/MeanWithinExpectedTest.cc]: Remove this macro. See comment above

File: source/src/plugins/MeanWithinExpectedTest.cc [hunk 9]
```diff
@@ -87,15 +95,19 @@ namespace dqm4hep {
...
+
+      SOFT_RETURN_RESULT_IF(XmlHelper::readParameter(xmlHandle, "MeanDeviationUpper", m_meanDeviationUpper));
+      
+      if ( (std::isnan(m_meanDeviationLower)) && (std::isnan(m_meanDeviationUpper)) ) {
```
Comment (rete) [inline source/src/plugins/MeanWithinExpectedTest.cc]: This test should goes in the `readParameter()` call as the validator lambda

File: source/src/plugins/MeanWithinExpectedTest.cc [hunk 10]
```diff
@@ -110,20 +122,64 @@ namespace dqm4hep {
...
 
-      if (m_meanDeviationLower < mean && mean < m_meanDeviationUpper) {
-        report.m_message =
-            "Within expected range: expected " + typeToString(m_expectedMean) + ", got " + typeToString(mean);
-      } 
+      // It's not relevant for now, since we're unable to do more than one test type per "qtest" object,
+      // since the tests only operate on one set of parameters.
+
+      for (std::vector<std::string>::iterator it = m_testTypes.begin(); it != m_testTypes.end(); it++){
+	result = AnalysisHelper::mainHelper(monitorElement, *it, m_percentage);
```
Comment (rete) [inline source/src/plugins/MeanWithinExpectedTest.cc]: You are not testing all tests here, but only the last one. The value of `result` is replaced at every iteration

File: source/src/plugins/MeanWithinExpectedTest.cc [hunk 11]
```diff
@@ -110,20 +122,64 @@ namespace dqm4hep {
...
+      }
+
+      if ( std::isnan(m_meanDeviationLower) ) {
```
Comment (rete) [inline source/src/plugins/MeanWithinExpectedTest.cc]: The full logic here has to be reviewed

**Review by rete — CHANGES_REQUESTED** [review]

File: source/include/dqm4hep/AnalysisHelper.h (hunk 1 above)
Comment (tcoates3) [reply source/include/dqm4hep/AnalysisHelper.h:49]: The reason I didn't define separate functions was to allow more flexibility, so you could make tests like mean75 or mean80, as well as mean90. If that's not desired, then I can revert to separate functions, which will allow me to remove the m_percentage variable and simplify some of the logic.

Comment (rete) [reply source/include/dqm4hep/AnalysisHelper.h:49]: Having a function like e.g `mean(element, 75)` is a nice thing to have. Could you just for convenience add one function mean90/rms90 as they are standard in HEP now simply call `mean(element, 90)` in the implementation.

About the string that is passed in the function mainHelper, I think the string parsing/evaluation should belong to the qtest itself and shouldn't appear here

File: source/include/dqm4hep/StatusCodes.h [hunk 12]
```diff
@@ -1,243 +0,0 @@
-/// \file StatusCodes.h
```
Comment (rete) [inline source/include/dqm4hep/StatusCodes.h]: Why did you removed this file ???

Comment (tcoates3) [reply source/include/dqm4hep/StatusCodes.h]: I was trying to remove the file's changes from the pull request and used "git rm" which removed the whole file! I'm not sure how to undo it.

File: source/src/plugins/MeanWithinExpectedTest.cc (hunk 7 above)
Comment (tcoates3) [reply source/src/plugins/MeanWithinExpectedTest.cc]: Do you mean m_percentage or m_testTypes?

Comment (rete) [reply source/src/plugins/MeanWithinExpectedTest.cc]: For m_testTypes

File: source/src/AnalysisHelper.cc [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+
+// -- dqm4hep headers
+#include <dqm4hep/AnalysisHelper.h>
+#include <typeinfo>
+
+namespace dqm4hep {
+
+  namespace core {
+
+    float AnalysisHelper::mean(MonitorElementPtr pMonitorElement, float percentage = 1.0)
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Please move the default value `= 1.0` to the header definition. Same for the other functions below.

File: source/src/AnalysisHelper.cc [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+    {
+
+      std::string ObjectType;
+
+      float result = 0.0;
+
+      TObject *ThisObject = pMonitorElement->objectTo<TObject>();
+
+      if (pMonitorElement->type() == "TH1F" || pMonitorElement->type() == "TH1I") {  // How many types of TH1 histogram do we have to recognise? TH1F, TH1I, ...?
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Here you can use something like : 
```cpp
if(nullptr != pMonitorElement->objectTo<TH1>() and nullptr == pMonitorElement->objectTo<TH2>) {
  // ...
}
```
Same for the other conditions below

File: source/src/AnalysisHelper.cc [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+	ObjectType = "TH1";
+      }
+      else if (pMonitorElement->type() == "TGraph") {
+	ObjectType = "TGraph";
+      }
+      else {
+	// Really this should be replaced with something that goes in the log
+	std::cout << "The monitor element " << pMonitorElement->name() << " of type " << pMonitorElement->type() << " could not be classified." << std::endl;
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Here use our logging interface : 
```cpp
dqm_error( "Message" );
```

File: source/src/AnalysisHelper.cc [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+	throw StatusCodeException(STATUS_CODE_FAILURE);
+      }
+
+      if (percentage == 1.0) {
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Never use equality operator for floats. Use numeric limits instead : 
```cpp
if(fabs(percentage - 1.f) < std::numeric_limits<float>::epsilon()) {

}
```

File: source/src/AnalysisHelper.cc [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+	      sumw += w;
+	      sumwx += w*x;
+	    }
+	    if (sumw > entries) break;
+	  }
+	  result = sumwx/sumw;
+	}
+
+	if (ObjectType == "TGraph") {
+	  TGraph *h = (TGraph*)(ThisObject);
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Here use the MonitorElement::objectTo<TGraph>() method

File: source/src/AnalysisHelper.cc [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+	  TAxis *axis = h->GetYaxis();
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: I'm not sure, but I remember by experience that the axis is not defined (nullptr) until you have drawn the graph itself. You might have a seg fault at runtime here

File: source/src/AnalysisHelper.cc [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+	  result = h->GetMean(1);
+	}
+	if (ObjectType == "TGraph") {
+	  TGraph *h = (TGraph*)(ThisObject);
+	  result = h->GetMean(2);
+	}
+      }
+      else {
+
+	if (ObjectType == "TH1") {
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: You can probably find a way to combine the two cases for TH1 and TGraph in a smart way.

File: source/src/AnalysisHelper.cc [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+	  }
+	  result = sumwx/sumw;
+	}
+      }
+      return result;
+    }
+
+    float AnalysisHelper::mean90(MonitorElementPtr pMonitorElement)
+    {
+      float result = AnalysisHelper::mean(pMonitorElement, 0.9);
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Just write : 
```cpp
return AnalysisHelper::mean(pMonitorElement, 0.9);
```
Same for the other functions below.

File: source/src/plugins/MeanWithinExpectedTest.cc [hunk 21]
```diff
@@ -87,43 +94,106 @@ namespace dqm4hep {
     //-------------------------------------------------------------------------------------------------
 
     StatusCode MeanWithinExpectedTest::readSettings(const TiXmlHandle xmlHandle) {
-      RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, XmlHelper::readParameter(xmlHandle, "ExpectedMean", m_expectedMean));
-
-      RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=,
-                       XmlHelper::readParameter(xmlHandle, "MeanDeviationLower", m_meanDeviationLower,
```
Comment (rete) [inline source/src/plugins/MeanWithinExpectedTest.cc]: Rename the parameter to DeviationLower as we allow for testing other properties. Same for DeviationUpper

File: source/src/plugins/MeanWithinExpectedTest.cc [hunk 22]
```diff
@@ -87,43 +94,106 @@ namespace dqm4hep {
...
+												       [this](const float &value){return !(isnan(this->m_meanDeviationLower) && isnan(value));}));
 
       return STATUS_CODE_SUCCESS;
     }
 
     //-------------------------------------------------------------------------------------------------
 
     void MeanWithinExpectedTest::userRun(MonitorElementPtr monitorElement, QualityTestReport &report) {
+
       TH1 *pHistogram = monitorElement->objectTo<TH1>();
```
Comment (rete) [inline source/src/plugins/MeanWithinExpectedTest.cc]: Should also work TGraph here !

File: source/src/plugins/MeanWithinExpectedTest.cc [hunk 23]
```diff
@@ -87,43 +94,106 @@ namespace dqm4hep {
...
+      {
+        // Do something
+      }
+      */
+
+      float result;
+
+      if (m_properties[0] == "Mean")
+	{
+	  result = AnalysisHelper::mean(monitorElement, m_percentage);
```
Comment (rete) [inline source/src/plugins/MeanWithinExpectedTest.cc]: You should not use the percentage here, as the user expected a "Mean" test. I thnik having Mean (without percentage), Mean90, RMS, RMS90 and Median is good start.

**Review by rete — CHANGES_REQUESTED** [review]

**Comment by rete** [comment]
More generally, about testing bounds with std::isnan, this can be replaced by using another parameter, which makes it more clear for the user. Add a new parameter "Method" with 3 choices : 
- LowerThan
- GreaterThan
- WithinRange

Then depending on the option you get in XmlHelper::readParameter(), you can parse the bounds properly. This method option can then be used in the qtest run method to take the appropriate action.

**Comment by rete** [comment]
About the unit test we talked about, you have an example here : 
https://github.com/DQM4HEP/dqm4hep-core/blob/master/source/tests/test-qtest-exact-ref-comp.cc
Please, write a new unit test in the same directory, it will be compile automatically (don't forget to re-run `cmake ..` to get it !).

File: source/src/AnalysisHelper.cc (hunk 15 above)
Comment (tcoates3) [reply source/src/AnalysisHelper.cc]: I'm trying to have the error message give both the incorrect argument *and* the name of the qtest where it occurred, doing it like this:
    dqm_error("{0} : Unrecognised test type: {1}", name, m_property);
and it gives me `error: invalid use of non-static member function`, so it seems like I'm using it incorrectly. Am I using the function wrong, or is it an issue with passing the `name` variable to the dqm_error macro?

Comment (rete) [reply source/src/AnalysisHelper.cc]: Yes, if you want to access the QTest name, you should use `name()` not `name`, which is, in the latter case, considered as a static member function (which is not, as the compiler says).

File: source/src/AnalysisHelper.cc (hunk 14 above)
Comment (tcoates3) [reply source/src/AnalysisHelper.cc]: Casting either a TGraph to a TH1, or a TH1 to a TGraph, both produce non-null pointers. This is what I have been using:

      if (nullptr != pMonitorElement->objectTo<TGraph>()) {
	ObjectType = "TGraph";
      }
      else if (nullptr != pMonitorElement->objectTo<TH1>()){
	ObjectType = "TH1";
      }
      else {
	dqm_error("The monitor element {0} of type {1} could not be recognised.", pMonitorElement->name(), pMonitorElement->type());
	throw StatusCodeException(STATUS_CODE_FAILURE);
      }

But this just results in the object being classified as whatever the first condition is, which then produces segfaults when I use a TH1-only method on a TGraph object and vice versa. In the meantime I've been adding a comparison of `pMonitorElement->type()` against a string, but is there a more foolproof way of determining the object type using casting? Can a TGraph be cast to a TH2?

Comment (rete) [reply source/src/AnalysisHelper.cc]: TGraph and TH1/TH2 are completely un-related. Casting one to another will raise a seg fault when using the casted-pointer. Look at the inheritance diagrams here : 
- TGraph : https://root.cern.ch/doc/master/classTGraph__inherit__graph_org.svg
- TH1 : https://root.cern.ch/doc/master/classTH1__inherit__graph_org.svg

Comment (tcoates3) [reply source/src/AnalysisHelper.cc]: For this purpose we're only interested in whether the cast itself returns a nullptr though, and it seems in any of these cases they do return a valid pointer, even if using it causes a segfault?

Comment (rete) [reply source/src/AnalysisHelper.cc]: I think I observed this problem at some point when implementing other QTest. I believe this was due to a wrong cast in the `MonitorElement::objectTo<T>` method. This is normally fixed in the head version. Have you already rebased your code on master ?

Comment (tcoates3) [reply source/src/AnalysisHelper.cc]: My branch isn't rebased to the master yet, I'll do that and see if the issue persists.

**Comment by tcoates3** [comment]
The last thing to do is code the unit test – I'm reading through the exact-ref-compare one and I want to be sure I'm understanding it correctly, so is there an example steering XML file for that test I can look at?

**Comment by rete** [comment]
That's the point. In these unit tests there is no corresponding XML file. The XML is created in memory in the source code. 
Example here : https://github.com/DQM4HEP/dqm4hep-core/blob/master/source/tests/test-qtest-exact-ref-comp.cc#L126

**Comment by tcoates3** [comment]
I know, but I wanted to see an XML file for the ExactRefCompare test so I can understand how to build the XML file in memory. The main thing I'm not understanding is that in the example you set parameters for the qtest using `qtestElement->SetAttribute("name", qtestName)` which is equivalent to the XML:

    <qtest type="PropertyWithinExpectedTest" name="MeanAround10Long" />

But I need to have parameters nested inside, e.g.:

    <qtest type="PropertyWithinExpectedTest" name="MeanAround10Long">
            <parameter name="ExpectedValue" value="10"/>
            <parameter name="DeviationLower" value="8"/>
            <parameter name="DeviationUpper" value="12"/>
            <parameter name="Property"> Mean </parameter>
    </qtest>

I can't find an equivalent function to use to set parameters within an XML element.

**Comment by rete** [comment]
You have a more complete example here : 
https://github.com/DQM4HEP/dqm4hep-core/blob/master/source/tests/test-qtest-fit-param-range.cc#L53
You can set parameters in the qtest element easily

**Comment by tcoates3** [comment]
So while I'm writing the unit test I'm trying to further error-proof the code, and I'm currently trying to use the XmlHelper function with validators in readSettings to make sure that when a LowerThan or HigherThan test is used without the corresponding bound, the program throws an exception immediately. I'm calling it like this:

    RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::readParameter(xmlHandle, "DeviationLower", m_deviationLower, [this](const float &value){return !(isnan(value) && (this->m_method=="HigherThan"));}));
    RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::readParameter(xmlHandle, "DeviationUpper", m_deviationUpper, [this](const float &value){return !(isnan(value) && (this->m_method=="LowerThan"));}));

But they never throw an exception! I've tested the delta functions used as the validators via couts, and they produce a 0 or 1 exactly as I intended. Am I using the validator/delta function incorrectly?

**Comment by rete** [comment]
Yes, sure the problem is the logic here. You ask your two lines if the method is `"HigherThan"` and `"LowerThan"`. In any case you get an exception. If I understand correctly, you should parse one parameter or the other depending on the method and not parse both parameters and validate it as a function of the method.

**Comment by tcoates3** [comment]
The thing that's confusing me is my logic statements are correct! They evaluate to 0 when doing a LowerThan test without a defined DeviationUpper, and 1 in all other cases. 

So I should replace it with something like this?:

    if (m_method == "HigherThan") {
        RETURN_RESULT_IF(STATUS_CODE_SUCCESS, != XmlHelper::readParameter(xmlHandle, "DevationLower", m_deviationLower));
    }
    else if (m_method == "LowerThan") {
         RETURN_RESULT_IF(STATUS_CODE_SUCCESS, != XmlHelper::readParameter(xmlHandle, "DevationUpper", m_deviationUpper));
    }
    else if (m_method == "WithinRange") {
        RETURN_RESULT_IF(STATUS_CODE_SUCCESS, != XmlHelper::readParameter(xmlHandle, "DevationUpper", m_deviationUpper));
        RETURN_RESULT_IF(STATUS_CODE_SUCCESS, != XmlHelper::readParameter(xmlHandle, "DevationLower", m_deviationLower));
    }
    else {
        dqm_error("Error message here");
        throw StatusCodeException(STATUS_CODE_FAILURE);
    }

I was hoping it would be a bit shorter in the validator functions...

**Comment by rete** [comment]
Sounds nice like this. Code length here is not a problem, it is just a few lines more and it's clearer.

**Comment by tcoates3** [comment]
Just as an FYI, I don't expect the current pushes to compile/work – currently rebasing the branch, then will solve some of the existing conflicts and changes from the updates to master, then finish validating the unit test. After that it should be ready to merge!

File: source/src/AnalysisHelper.cc [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+      }
+
+      if(fabs(percentage - 1.f) < std::numeric_limits<float>::epsilon()) {
+	if (ObjectType == "TH1") {
+	  TH1 *h = pMonitorElement->objectTo<TH1>();
+	  result = h->GetMean(1);
+	}
+	if (ObjectType == "TGraph") {
+	  TGraph *h = pMonitorElement->objectTo<TGraph>();
+	  result = h->GetMean(1);
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Do you mean `GetMean(2)` here ?

File: source/src/AnalysisHelper.cc [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+    float AnalysisHelper::mean(MonitorElement* pMonitorElement, float percentage)
+    {
+
+      std::string ObjectType;
+
+      float result = 0.0;
+
+      //TObject *ThisObject = pMonitorElement->objectTo<TObject>();
+
+      if (nullptr != pMonitorElement->objectTo<TGraph>() && pMonitorElement->type() == "TGraph") {
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Double check for nothing here. Just `objectTo<TGraph>` is enough

File: source/src/AnalysisHelper.cc [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+	    if (sumw > entries) break;
+	  }
+	  result = sumwx/sumw;
+	}
+
+	if (ObjectType == "TGraph") {
+	  //TGraph *h = (TGraph*)(ThisObject);
+	  TGraph *h = pMonitorElement->objectTo<TGraph>();
+
+	  double* arrayX = h->GetX();
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Same as above, do you mean GetY() instead ?

File: source/src/AnalysisHelper.cc [hunk 27]
```diff
@@ -0,0 +1,244 @@
...
+
+    float AnalysisHelper::rms(MonitorElement* pMonitorElement, float percentage)
+    {
+      std::string ObjectType;
+
+      float result = 0.0;
+
+      //TObject *ThisObject = pMonitorElement->objectTo<TObject>();
+
+      if (nullptr != pMonitorElement->objectTo<TGraph>() && pMonitorElement->type() == "TGraph") {
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Unneeded double check here also

File: source/src/AnalysisHelper.cc [hunk 28]
```diff
@@ -0,0 +1,244 @@
...
+	    if (sumw > entries) break;
+	  }
+	  result = pow(sumwx/sumw,0.5);
+	}
+	if (ObjectType == "TGraph") {
+	  percentage = 0.9;
+	  //TGraph *h = (TGraph*)(ThisObject);
+	  TGraph *h = pMonitorElement->objectTo<TGraph>();
+
+	  double* arrayX = h->GetX();
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Same here, do you mean `GetY()` ?

File: source/src/AnalysisHelper.cc [hunk 29]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    float AnalysisHelper::rms90(MonitorElement* pMonitorElement)
+    {
+      return AnalysisHelper::rms(pMonitorElement, 0.9);
+    }
+
+    float AnalysisHelper::median(MonitorElement* pMonitorElement)
+    {
+      // This method works ONLY for a TH1 and must be changed to handle a TGraph
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: TGraph option ?

File: source/src/plugins/MeanWithinExpectedTest.cc [hunk 30]
```diff
@@ -33,6 +33,7 @@
 #include <dqm4hep/QualityTest.h>
```
Comment (rete) [inline source/src/plugins/MeanWithinExpectedTest.cc]: You have now the PropertyWithinExpected.cc file instead, please remove this one.

File: source/src/plugins/PropertyWithinExpectedTest.cc [hunk 31]
```diff
@@ -0,0 +1,219 @@
...
+    PropertyWithinExpectedTest::PropertyWithinExpectedTest(const std::string &qname)
+        : QualityTest("MeanWithinExpected", qname),
+          m_expectedValue(0.f),
+	  m_deviationLower(float NAN),
+	  m_deviationUpper(float NAN),
+	  m_percentage(1.0),
+	  m_property(),
+	  m_method("WithinRange")
+{
+      m_description = "Test if the mean of the histogram if contained in the expected user range. The quality is "
```
Comment (rete) [inline source/src/plugins/PropertyWithinExpectedTest.cc]: Update your description here. It is not what the test is doing

File: source/src/plugins/PropertyWithinExpectedTest.cc [hunk 32]
```diff
@@ -0,0 +1,219 @@
...
+      return std::shared_ptr<QTest>(new PropertyWithinExpectedTest(qname));
+    }
+
+    //-------------------------------------------------------------------------------------------------
+    //-------------------------------------------------------------------------------------------------
+
+    PropertyWithinExpectedTest::PropertyWithinExpectedTest(const std::string &qname)
+        : QualityTest("MeanWithinExpected", qname),
+          m_expectedValue(0.f),
+	  m_deviationLower(float NAN),
```
Comment (rete) [inline source/src/plugins/PropertyWithinExpectedTest.cc]: Set it to `0.f` here, both `m_deviationLower` and `m_deviationUpper`

**Review by rete — CHANGES_REQUESTED** [review]

File: source/src/AnalysisHelper.cc [hunk 33]
```diff
@@ -0,0 +1,244 @@
...
+	  }
+	  result = sumwx/sumw;
+	}
+
+	if (ObjectType == "TGraph") {
+	  //TGraph *h = (TGraph*)(ThisObject);
+	  TGraph *h = pMonitorElement->objectTo<TGraph>();
+
+	  double* arrayX = h->GetX();
+	  int entries = h->GetHistogram()->GetNbinsX();
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Please use `GetN()` instead. Don't use `GetHistogram()` anywhere

**Review by rete — CHANGES_REQUESTED** [review]

**Comment by tcoates3** [comment]
@rete  The [CI test](https://travis-ci.org/DQM4HEP/dqm4hep-core/jobs/385286847) has failed because `test-quality-test-test` failed. The new unit test I implemented (`test-qtest-property-within-exp-test`) passed this check. Have I inadvertently broken one of the other unit tests, or is something else going wrong?

**Comment by rete** [comment]
You have to run your new GenerateData.C to get a root file with your TGraphs. You did this locally, but you have not pushed the root file. The unit test then can't find your TGraphs. Please push also the root file. It should make the trick.

**Comment by tcoates3** [comment]
@rete There's still some requested changes pending, but I think I've addressed all of them. I feel like this is complete now, and the Travis CI tests have all passed.

File: source/src/AnalysisHelper.cc [hunk 34]
```diff
@@ -0,0 +1,215 @@
...
+#include <typeinfo>
+#include <TH2.h>
+
+namespace dqm4hep {
+
+  namespace core {
+
+    float AnalysisHelper::mean(MonitorElement* pMonitorElement, float percentage)
+    {
+      TH1 *pTH1 = new TH1F; // While this is declared as a TH1F, it doesn't seem to matter since it's overwritten below. It's left as a float as this is the most likely type.
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Why are you creating a new TH1 here ? You have a memory leak.

Comment (rete) [reply source/src/AnalysisHelper.cc]: Same for TGraph below

File: source/src/AnalysisHelper.cc [hunk 35]
```diff
@@ -0,0 +1,215 @@
...
+    }
+
+    float AnalysisHelper::mean90(MonitorElement* pMonitorElement)
+    {
+      return AnalysisHelper::mean(pMonitorElement, 0.9);
+    }
+
+    float AnalysisHelper::rms(MonitorElement* pMonitorElement, float percentage)
+    {
+      TH1 *pTH1 = new TH1F;
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: Same as above, you have memory leaks here !

File: source/src/AnalysisHelper.cc [hunk 36]
```diff
@@ -0,0 +1,215 @@
...
+
+      if (nullptr != pMonitorElement->objectTo<TH1>()) {
+	Double_t xq[1];
+	Double_t yq[1];
+	xq[0] = 0.5;
+	pMonitorElement->objectTo<TH1>()->GetQuantiles(1, yq, xq);
+	return yq[0];
+      }
+      else if (nullptr != pMonitorElement->objectTo<TGraph>()) {
+	double* arrayY = pMonitorElement->objectTo<TGraph>()->GetY();
```
Comment (rete) [inline source/src/AnalysisHelper.cc]: For empty TGraphs you have a problem in the following

**Review by rete — CHANGES_REQUESTED** [review]

**Comment by rete** [comment]
Please, rebase before I merge

**Review by rete — APPROVED** [review]

**Pull request merged**